### PR TITLE
Add PlayerGlassBottleFillEvent

### PR DESCRIPTION
**Pull request by JonMagon**
I hope it would be helpful, the event is fired when the bottle is filled:
- via a water source
- via a cauldron
- from a beehive
- with dragon's breath.

Doing the same with PlayerInteractEvent in most cases, for example, as replacing with other ItemStack, is impossible or really hard.

### CODE REVIEW

File: Spigot-API-Patches/0274-Add-PlayerGlassBottleFillEvent.patch [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
++    /**
++     * Gets the existing item that's being replaced. Modifying the returned item will
++     * have no effect, you should use {@link
++     * #setNewItem(ItemStack)} instead.
++     *
++     * @return old item
++     */
++    @NotNull
++    public ItemStack getItem() {
++        return item.clone();
```
Comment (JRoy) [inline Spigot-API-Patches/0274-Add-PlayerGlassBottleFillEvent.patch]: Any reason why you're cloning this? What if someone wants to cancel the event but modify the original bottle?

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
++import io.papermc.paper.event.player.PlayerGlassBottleFillEvent; // Paper
+ 
+ import java.util.Iterator;
+ import java.util.List;
+@@ -94,13 +95,30 @@ public class BlockBeehive extends BlockTileEntity {
+                 });
+                 flag = true;
+             } else if (itemstack.getItem() == Items.GLASS_BOTTLE) {
++                // Paper start - PlayerGlassBottleFillEvent
++                PlayerGlassBottleFillEvent event = new PlayerGlassBottleFillEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(new ItemStack(Items.HONEY_BOTTLE)));
```
Comment (JRoy) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: `asBukkitCopy` -> `asCraftMirror` as per my last comment

```suggestion
+                PlayerGlassBottleFillEvent event = new PlayerGlassBottleFillEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(new ItemStack(Items.HONEY_BOTTLE)));
```

Comment (JRoy) [reply Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: Also, for the HoneyBottle, just create a bukkit itemstack, no point in making a NMS one if you're going to copy it anyway

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
++                if (!event.callEvent()) {
++                    if (entityhuman instanceof EntityPlayer) {
++                        ((EntityPlayer) entityhuman).updateInventory(entityhuman.defaultContainer);
```
Comment (JRoy) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: This is up to plugin implementation, you shouldn't be doing this here. Add a note like I did [here](https://github.com/PaperMC/Paper/blob/master/Spigot-API-Patches/0251-Add-EntityLoadCrossbowEvent.patch#L88-L97) to the `setCanceled` method of your event

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+-                    entityhuman.drop(new ItemStack(Items.HONEY_BOTTLE), false);
++                    entityhuman.a(enumhand, newitem);
++                } else if (!entityhuman.inventory.pickup(newitem)) {
++                    entityhuman.drop(newitem, false);
++                }
++                // Paper start - PlayerGlassBottleFillEvent, fix visual glitch
++                else if (entityhuman instanceof EntityPlayer) {
++                    ((EntityPlayer) entityhuman).updateInventory(entityhuman.defaultContainer);
+                 }
++                // Paper end - PlayerGlassBottleFillEvent
```
Comment (JRoy) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: see above

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+@@ -100,6 +101,16 @@ public class BlockCauldron extends Block {
+                         }
+                         if (!entityhuman.abilities.canInstantlyBuild) {
+                             itemstack1 = PotionUtil.a(new ItemStack(Items.POTION), Potions.WATER);
++                            // Paper start - PlayerGlassBottleFillEvent
++                            PlayerGlassBottleFillEvent event = new PlayerGlassBottleFillEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack1));
++                            if (!event.callEvent()) {
++                                if (entityhuman instanceof EntityPlayer) {
++                                    ((EntityPlayer) entityhuman).updateInventory(entityhuman.defaultContainer);
++                                }
```
Comment (JRoy) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: above

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
++    // Paper start - PlayerGlassBottleFillEvent
++    protected InteractionResultWrapper<ItemStack> handleInteractedItemStack(World world, EntityHuman entityhuman, ItemStack itemstack, ItemStack newitemstack) {
++        PlayerGlassBottleFillEvent event = new PlayerGlassBottleFillEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(itemstack), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(newitemstack));
++        if (event.callEvent()) {
++            return InteractionResultWrapper.a(this.a(itemstack, entityhuman, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getNewItem())), world.s_());
++        }
++        else {
++            if (entityhuman instanceof EntityPlayer) {
++                ((EntityPlayer) entityhuman).updateInventory(entityhuman.defaultContainer);
++            }
```
Comment (JRoy) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch]: above

**Review by JRoy — CHANGES_REQUESTED** [review]
see comments

File: Spigot-API-Patches/0274-Add-PlayerGlassBottleFillEvent.patch (hunk 1 above)
Comment (JonMagon) [reply Spigot-API-Patches/0274-Add-PlayerGlassBottleFillEvent.patch]: Thank you, I forgot about it, I thought I would need to use getItem() in the patch

**Comment by JonMagon** [comment]
Thanks

I don't know how, but I removed the request for review above. 😳

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+ 
+@@ -45,6 +53,18 @@ public class ItemGlassBottle extends Item {
+         }
+     }
+ 
++    // Paper start - PlayerGlassBottleFillEvent
++    protected InteractionResultWrapper<ItemStack> handleInteractedItemStack(World world, EntityHuman entityhuman, ItemStack itemstack, ItemStack newitemstack) {
++        PlayerGlassBottleFillEvent event = new PlayerGlassBottleFillEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), org.bukkit.craftbukkit.inventory.CraftItemStack.asCraftMirror(itemstack), org.bukkit.craftbukkit.inventory.CraftItemStack.asBukkitCopy(newitemstack));
++        if (event.callEvent()) {
++            return InteractionResultWrapper.a(this.a(itemstack, entityhuman, org.bukkit.craftbukkit.inventory.CraftItemStack.asNMSCopy(event.getNewItem())), world.s_());
```
Comment (MiniDigger) [inline Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch:114]: add a note here that this line was copied from above. I think it doesnt need obf helpers then, as its easy to find the new names by just copying the line again

**Review by MiniDigger — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch (hunk 7 above)
Comment (JonMagon) [reply Spigot-Server-Patches/0673-Add-PlayerGlassBottleFillEvent.patch:114]: I've added. Is it right?

**Comment by JonMagon** [comment]
I'll open a new PR, there's something confusing about commits in master.

**Pull request closed without merging**